File: 2020/Javascript/d9/p2.js
let fs = require('fs');

function sol(data) {
    data = data.map(x => parseInt(x));
    const TARGET = 542529149;

    let f = () => {
        for (let n = 2; n < data.length; n++) {
            let total = data.slice(0, n);
            for (let i = n; i < data.length; i++) {
                if (sum(total) == TARGET) {
                    //console.log(total);
                    return total;
                }
                total.shift();
                total.push(data[i]);
            }
        }
    };

    let ans = f();
    return Math.min(...ans) + Math.max(...ans);
}

function sum(lst) {
    return lst.reduce((a, b) => a + b, 0);
}

function getPossible(curList) {
    let f = {};
    curList.forEach((a, i) => {
        curList.forEach((b, j) => {
            if (i != j) {
                f[a + b] = true;
            }
        });
    });
    return f;
}

//
// main
//

async function main() {
    let data;
    try {
        data = await readFile('./input.txt');
    } catch(err) { throw err; }

    data = data.trim().split("\n");
    let ans = sol(data);

    if(ans !== undefined) {
        console.log(ans);
    }
}

main().catch(err => console.log(err));

//
// Helper Functions
//

// read file
function readFile(path) {
    return new Promise((res, rej) => {
        fs.readFile(path, 'utf8', (err, data) => {
            if (err) rej(err);
            res(data);
        });
    });
}

File: 2019/Python/D16/p2.py
import sys
lst = [int(x) for x in list(open("input.txt").read()[:-1])]
lst = lst * 10000
D = False

def fft(lst):
    outlst = []
    total = sum(lst)
    for i in range(len(lst)):
        outlst.append(abs(total) % 10)
        total -= lst[i]
    return outlst

def sol(lst):
    baseline = int("".join([str(x) for x in lst[:7]]))
    print("baseline:", baseline)
    lst = lst[baseline:]
    for i in range(100):
        if D: print("lst:", lst)
        lst = fft(lst)
        print(i)
        sys.stdout.flush()
    if D: print("lst:", lst)
    return lst[:8]

ans = sol(lst)
print(ans)
